### PR TITLE
Postgres Vanilla Testing Correctness with Citus Extension

### PR DESCRIPTION
**Pull request by aykut-bozkurt**
* We ensure citus resides on preloaded shared libraries on nodes before tests start.
* We ensure citus does not break postgres vanilla tests. There are 12 subtasks which are described in the checklist below. You can see them to look at what we have done so far to be compatible with pg tests without changing citus behaviour as much as possible.

1. #6055,
2. #6063,
3. #6056,
4. #6088 ,
5. #6081,
6. #6082,
7. #6120,
8.  #6125,
9. #6083,
10. #6131,
11. #6132,
12. #6130.

### CODE REVIEW

File: src/test/regress/pg_regress_multi.pl [hunk 1]
```diff
@@ -920,56 +920,72 @@ END
         system(catfile($bindir, "psql"),
             ('-X', '-h', $host, '-p', $port, '-U', $user, "-d", "postgres",
                 '-c', "CREATE DATABASE regression;")) == 0
-            or die "Could not create regression database on worker";
+            or die "Could not create regression database on worker port $port.";
+        print "Created database regression on worker port $port. \n";
+        system(catfile($bindir, "psql"),
```
Comment (onderkalaci) [inline src/test/regress/pg_regress_multi.pl:925]: I think showing alone could be fragile. What if this behavior changes? We cannot automatically detect.

Instead, can we maybe change the script such that if `shared_preload_libraries` does not include `citus`, we go to `die` instead of just printing it?

**Review by onderkalaci — CHANGES_REQUESTED** [review]
Unfortunately we are not creating the extension on the coordinator. If you check the logs, you'd see we are only creating on the workers:
```
CREATE EXTENSION
Created extension citus on worker port 57637. 
CREATE FUNCTION
Created function fake_fdw_handler() on worker port 57637. 
CREATE FOREIGN DATA WRAPPER
Created foreign data wrapper fake_fdw on worker port 57637. 
CREATE DATABASE
Created database regression on worker port 57638. 
```

where the extension is only created on the workers. However, check-vanilla runs on coordinator, port `57636`.

While running the `check-vanilla` locally, you can connect to the database and see that the extension is not created:

`psql -p 57636 regression -U postgres -h 127.0.0.1`

```SQL

 \dx
                  List of installed extensions
┌─────────┬─────────┬────────────┬──────────────────────────────┐
│  Name   │ Version │   Schema   │         Description          │
├─────────┼─────────┼────────────┼──────────────────────────────┤
│ plpgsql │ 1.0     │ pg_catalog │ PL/pgSQL procedural language │
└─────────┴─────────┴────────────┴──────────────────────────────┘
(1 row)
```

I think the `--load-extension=` is a confusing name, it only creates the extensions on the workers. One way could be we extend that parameter to create the extensions on the coordinator as well.

**Comment by onderkalaci** [comment]
Btw, when I create the extension at the start of the check-vanilla manually, I can see 47 tests fail:

```
=========================
 47 of 210 tests failed. 
=========================
```

Majority of the failures are these types of Citus warnings:
```
+WARNING:  "function base_fn_out(base_type)" has dependency on unsupported object "type base_type"
+DETAIL:  "function base_fn_out(base_type)" will be created only locally
``` 

By adding a GUC, and somehow making it NOT print these warnings coulld drop the number of failed tests to ~10-~15.

For others, we may need to look one-by-one

File: src/test/regress/pg_regress_multi.pl [hunk 2]
```diff
@@ -920,56 +920,74 @@ END
...
         }
 
         foreach my $function (keys %functions)
         {
             system(catfile($bindir, "psql"),
                     ('-X', '-h', $host, '-p', $port, '-U', $user, "-d", "regression",
                     '-c', "CREATE FUNCTION $function RETURNS $functions{$function};")) == 0
-                or die "Could not create FUNCTION $function on worker";
+                or die "Could not create function $function on worker port $port";
+            print "Created function $function on worker port $port. \n";
```
Comment (onderkalaci) [inline src/test/regress/pg_regress_multi.pl:946]: I think I'd remove the new `print`s added in this PR, they create some noise?

**Comment by aykut-bozkurt** [comment]
- [ ] Avoid WARNINGs via GUC,
- [ ] Qualified object names in the error message (can we use normalization rules? Or, fix in Citus),
- [ ] Find a way to fix that PG query shows Citus objects,
- [ ] Fix conflicting relation/table/sequence error messages,
- [ ] Open issues for citus bugs (DROP VIEW cascade, DROP TEXT SEARCH IF EXISTS).

File: src/backend/distributed/utils/meta_visibility.c [hunk 3]
```diff
@@ -0,0 +1,700 @@
...
+
+PG_FUNCTION_INFO_V1(is_citus_depended_object);
+
+/*
+ * is_citus_depended_object a wrapper around IsCitusDependentObject, so
+ * see the details there.
+ */
+Datum
+is_citus_depended_object(PG_FUNCTION_ARGS)
+{
```
Comment (onderkalaci) [inline src/backend/distributed/utils/meta_visibility.c:64]: can we use `IsObjectAddressOwnedByExtension` function instead of creating a new one? Maybe we create a variation of it it to `IsObjectAddressOwnedByCitusExtension` by calling `IsObjectAddressOwnedByExtension`?

That could help us to remove bunch of code here?

File: src/backend/distributed/utils/meta_visibility.c [hunk 4]
```diff
@@ -0,0 +1,700 @@
...
+			CurrentMemoryContext,
+			"Dependent Objects Context",
+			ALLOCSET_DEFAULT_SIZES);
+
+	MemoryContext oldContext = MemoryContextSwitchTo(DependentObjectsContext);
+
+	switch (metaTableId)
+	{
+		/* handle meta objects that can be found in pg_depend */
+		case RelationRelationId:
```
Comment (onderkalaci) [inline src/backend/distributed/utils/meta_visibility.c:89]: In general, this code seems to be specific per object type.

Is there a possibility to use `PgGetObjectAddress()` function such that we have a generic implementation?

File: src/backend/distributed/shared_library_init.c [hunk 5]
```diff
@@ -848,6 +849,16 @@ RegisterCitusConfigVariables(void)
 		GUC_STANDARD,
 		NULL, NULL, NULL);
 
+	DefineCustomBoolVariable(
```
Comment (onderkalaci) [inline src/backend/distributed/shared_library_init.c:852]: I think we should find a more specific name than `disable_citus_preconditions` but I'll think about that later.

Do you have any other suggestions?

File: src/backend/distributed/citus.control [hunk 6]
```diff
@@ -1,6 +1,6 @@
 # Citus extension
 comment = 'Citus distributed database'
-default_version = '11.1-1'
+default_version = '11.1-2'
```
Comment (marcocitus) [inline src/backend/distributed/citus.control:3]: we only use one upgrade script until we cut a release, so this should be 11.1-1

File: src/backend/distributed/commands/grant.c [hunk 7]
```diff
@@ -91,9 +91,14 @@ PreprocessGrantStmt(Node *node, const char *queryString,
 
 			if (priv->cols != NIL)
 			{
-				ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-								errmsg("grant/revoke on column list is currently "
-									   "unsupported")));
+				if (!DisablePreconditions)
```
Comment (marcocitus) [inline src/backend/distributed/commands/grant.c:94]: if this affects non-Citus tables it seems like a bug

File: src/backend/distributed/commands/cluster.c [hunk 8]
```diff
@@ -40,9 +40,12 @@ PreprocessClusterStmt(Node *node, const char *clusterCommand,
 
 	if (clusterStmt->relation == NULL)
 	{
-		ereport(WARNING, (errmsg("not propagating CLUSTER command to worker nodes"),
-						  errhint("Provide a specific table in order to CLUSTER "
-								  "distributed tables.")));
+		if (!DisablePreconditions)
```
Comment (marcocitus) [inline src/backend/distributed/commands/cluster.c:43]: maybe `EnablePropagationWarnings`?

(and use separate GUCs for other purposes)

File: src/backend/distributed/commands/grant.c (hunk 7 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/grant.c:94]: Yes, makse sense, we should be careful for such cases.

For example, we can check if `CollectGrantTableIdList` returns any tables that are in Citus' metadata such that `RelationIdListContainsCitusTableType(tableIdList, ANY_CITUS_TABLE_TYPE)` or such

Comment (aykut-bozkurt) [reply src/backend/distributed/commands/grant.c:94]: I removed the GUC, and it does not cause a failure. I guess I left that behind while I am testing something at start.

Comment (aykut-bozkurt) [reply src/backend/distributed/commands/grant.c]: Added more GUCs for different tasks related to vanilla testing. And renamed the old GUC.

File: src/backend/distributed/commands/cluster.c (hunk 8 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/commands/cluster.c]: Added more GUCs for different tasks related to vanilla testing. And renamed the old GUC.

File: src/backend/distributed/commands/index.c [hunk 9]
```diff
@@ -321,15 +321,25 @@ ExecuteFunctionOnEachTableIndex(Oid relationId, PGIndexProcessor pgIndexProcesso
 	List *result = NIL;
 
 	Relation relation = RelationIdGetRelation(relationId);
+	if (!RelationIsValid(relation))
```
[the remark this answers is not in the document]
Comment (aykut-bozkurt) [reply src/backend/distributed/commands/index.c]: Opened PR #6048 for that.

File: src/backend/distributed/deparser/deparse_view_stmts.c [hunk 10]
```diff
@@ -157,8 +158,13 @@ AppendAlterViewCmd(StringInfo buf, AlterTableCmd *alterTableCmd)
 
 		case AT_ColumnDefault:
 		{
-			elog(ERROR, "Citus doesn't support setting or resetting default values for a "
-						"column of view");
+			if (EnablePropagationWarnings)
```
Comment (onderkalaci) [inline src/backend/distributed/deparser/deparse_view_stmts.c:161]: why do we ever need to deparse a non-distributed view such that we have to change `AppendAlterViewCmd`?

File: src/backend/distributed/commands/grant.c [hunk 11]
```diff
@@ -94,6 +94,8 @@ PreprocessGrantStmt(Node *node, const char *queryString,
 				ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
 								errmsg("grant/revoke on column list is currently "
 									   "unsupported")));
+
+				return NIL;
```
Comment (onderkalaci) [inline src/backend/distributed/commands/grant.c:98]: It is redundant to have return after ERROR, because we'd already return at that line

File: src/backend/distributed/commands/index.c [hunk 12]
```diff
@@ -328,8 +328,13 @@ ExecuteFunctionOnEachTableIndex(Oid relationId, PGIndexProcessor pgIndexProcesso
 		HeapTuple indexTuple = SearchSysCache1(INDEXRELID, ObjectIdGetDatum(indexId));
 		if (!HeapTupleIsValid(indexTuple))
 		{
-			ereport(ERROR, (errmsg("cache lookup failed for index with oid %u",
-								   indexId)));
+			if (EnablePropagationWarnings)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/index.c:331]: This is also changing behavior. Can you remind which test does it fix?

(I think the name of the GUC is relatively good, we should only change behavior for WARNINGs, not ERRORs. Because errors are changing behavior)

File: src/backend/distributed/commands/index.c [hunk 13]
```diff
@@ -556,16 +561,26 @@ PreprocessReindexStmt(Node *node, const char *reindexCommand,
 			state.locked_table_oid = InvalidOid;
 
 			Oid indOid = RangeVarGetRelidExtended(reindexStatement->relation,
-												  lockmode, 0,
+												  lockmode, RVR_MISSING_OK,
 												  RangeVarCallbackForReindexIndex,
 												  &state);
+			if (!OidIsValid(indOid))
+			{
+				return NIL;
```
Comment (onderkalaci) [inline src/backend/distributed/commands/index.c:569]: again, lets add 1 line comments like `/* let Postgres decide on the behavior and the error message */`

File: src/backend/distributed/commands/table.c [hunk 14]
```diff
@@ -439,9 +441,12 @@ PreprocessAlterTableStmtAttachPartition(AlterTableStmt *alterTableStatement,
 			Oid parentRelationId = AlterTableLookupRelation(alterTableStatement,
 															lockmode);
 			PartitionCmd *partitionCommand = (PartitionCmd *) alterTableCommand->def;
-			bool partitionMissingOk = false;
 			Oid partitionRelationId = RangeVarGetRelid(partitionCommand->name, lockmode,
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:444]: it is good to have `partitionMissingOk` variable, which makes the code more readable.

File: src/backend/distributed/commands/text_search.c [hunk 15]
```diff
@@ -710,16 +710,19 @@ AlterTextSearchConfigurationSchemaStmtObjectAddress(Node *node, bool missing_ok)
...
-			 * if the text search config id is still invalid we couldn't find it, error
-			 * with the same message postgres would error with if missing_ok is false
-			 * (not ok to miss)
-			 */
-
-			ereport(ERROR,
-					(errcode(ERRCODE_UNDEFINED_OBJECT),
-					 errmsg("text search configuration \"%s\" does not exist",
-							NameListToString(castNode(List, stmt->object)))));
+			if (EnablePropagationWarnings)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/text_search.c:713]: Again, we should try to avoid masking an error, this might be a Citus bug. 

Can this function return an InvalidObjectAddr, and let the caller return early and PG to give the proper error?

File: src/backend/distributed/commands/utility_hook.c [hunk 16]
```diff
@@ -706,10 +706,16 @@ ProcessUtilityInternal(PlannedStmt *pstmt,
...
+			if (EnablePropagationWarnings)
+			{
+				ereport(NOTICE, (errmsg(
+									 "not propagating ALTER ROLE ... RENAME TO commands "
+									 "to worker nodes"),
+								 errhint("Connect to worker nodes directly to manually "
+										 "rename the role")));
+			}
+
+			return;
```
Comment (onderkalaci) [inline src/backend/distributed/commands/utility_hook.c:718]: I don't think we are supposed to return here. Is that really needed?

File: src/backend/distributed/deparser/citus_ruleutils.c [hunk 17]
```diff
@@ -1039,7 +1039,12 @@ pg_get_indexclusterdef_string(Oid indexRelationId)
 										  0, 0, 0);
 	if (!HeapTupleIsValid(indexTuple))
 	{
-		ereport(ERROR, (errmsg("cache lookup failed for index %u", indexRelationId)));
+		if (EnablePropagationWarnings)
```
Comment (onderkalaci) [inline src/backend/distributed/deparser/citus_ruleutils.c:1042]: `pg_get_indexclusterdef_string` is only called for `deparsing` queries. 

How a PG index comes here? We should most likely prevented the code flow coming here instead of masking an ERROR here?

File: src/backend/distributed/commands/utility_hook.c (hunk 16 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/commands/utility_hook.c:718]: No, it is not, I removed it.

File: src/backend/distributed/deparser/deparse_view_stmts.c (hunk 10 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/deparser/deparse_view_stmts.c:161]: It is also left behind from start. Tests pass without it, so I removed the check.

File: src/backend/distributed/commands/index.c (hunk 12 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/commands/index.c:331]: It is left from the beginning of the PR. It is unnecessary again.

File: src/backend/distributed/deparser/citus_ruleutils.c (hunk 17 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/deparser/citus_ruleutils.c:1042]: This is also unnecessary check. Tests pass without it.

File: src/backend/distributed/commands/text_search.c (hunk 15 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/commands/text_search.c:713]: It is left from the beginning of the PR. It is unnecessary again.

**Comment by aykut-bozkurt** [comment]
Besides from the ones in comments, there were some more unnecessary EnablePropagationWarnings checks which does not affect postgres tests. I removed them.

File: src/backend/distributed/commands/function.c [hunk 18]
```diff
@@ -1366,6 +1366,11 @@ PreprocessCreateFunctionStmt(Node *node, const char *queryString,
 List *
 PostprocessCreateFunctionStmt(Node *node, const char *queryString)
 {
+	if (!EnablePropagationWarnings)
```
[the remark this answers is not in the document]
Comment (aykut-bozkurt) [reply src/backend/distributed/commands/function.c:1369]: We have 2 such cases (QualifyAlterDomainStmt and PostprocessCreateFunctionStmt) in which we change behaviour because they cause postgres test failures. I need to investigate those.

File: src/backend/distributed/commands/sequence.c [hunk 19]
```diff
@@ -266,7 +266,16 @@ PreprocessDropSequenceStmt(Node *node, const char *queryString,
 	{
 		RangeVar *seq = makeRangeVarFromNameList(objectNameList);
 
-		Oid seqOid = RangeVarGetRelid(seq, NoLock, stmt->missing_ok);
+		Oid seqOid = RangeVarGetRelid(seq, NoLock, true);
+
+		if (!stmt->missing_ok && !OidIsValid(seqOid))
+		{
+			/*
```
Comment (onderkalaci) [inline src/backend/distributed/commands/sequence.c:273]: Do we really need to check `stmt->missing_ok` here? I mean, as long as `!OidIsValid(seqOid)` we should just rely on Postgres, shouldn't we?

File: src/backend/distributed/commands/index.c [hunk 20]
```diff
@@ -556,16 +556,34 @@ PreprocessReindexStmt(Node *node, const char *reindexCommand,
 			state.locked_table_oid = InvalidOid;
 
 			Oid indOid = RangeVarGetRelidExtended(reindexStatement->relation,
-												  lockmode, 0,
+												  lockmode, RVR_MISSING_OK,
 												  RangeVarCallbackForReindexIndex,
 												  &state);
+			if (!OidIsValid(indOid))
```
Comment (onderkalaci) [inline src/backend/distributed/commands/index.c:562]: I think we can extract this class of changes to another PR as well: Properly check existence of objects in Preprocess functions

File: src/backend/distributed/deparser/deparse_view_stmts.c [hunk 21]
```diff
@@ -17,6 +17,7 @@
 #include "distributed/commands.h"
 #include "distributed/deparser.h"
 #include "distributed/listutils.h"
+#include "distributed/log_utils.h"
```
Comment (onderkalaci) [inline src/backend/distributed/deparser/deparse_view_stmts.c:20]: can you undo the changes in this file? That is mostly to keep the git history clean

File: src/backend/distributed/deparser/qualify_statistics_stmt.c [hunk 22]
```diff
@@ -83,6 +86,12 @@ QualifyDropStatisticsStmt(Node *node)
 
 			if (OidIsValid(statsOid))
 			{
+				/* user should own the statistic object */
```
Comment (onderkalaci) [inline src/backend/distributed/deparser/qualify_statistics_stmt.c:89]: This is way too low in the call stack to check the owner of the object. Is it possible to move this check to earlier?

File: src/backend/distributed/deparser/qualify_text_search_stmts.c [hunk 23]
```diff
@@ -92,7 +101,16 @@ QualifyDropTextSearchDictionaryStmt(Node *node)
 
 		if (!schemaName)
 		{
-			Oid tsdictOid = get_ts_dict_oid(objName, false);
+			Oid tsdictOid = get_ts_dict_oid(objName, true);
```
Comment (onderkalaci) [inline src/backend/distributed/deparser/qualify_text_search_stmts.c:104]: I think deciding anything in a `Qualify*` function is really hard to follow, and probably not right. So, we return unqualified version of the input, which is NOT expected.  

One idea I'm tinkering is whether we should add `exists()` function to `DistributeObjectOps`. 

With that, we can do something like below in the pre-process:

```
if (ops && ops->exists)
{
    objectExists = ops->exists(parsetree)
}

if (objectExists)
{
   .. rest of the qualify, preprocess is called
}
```

With that, we'd avoid lots of code changes in this PR, and prevent any future such bugs. If the object does not exist, we just let Postgres do its job.

For now, the default/generic `exists` function can blindly return `true`, and for the objects that fails in check-vanilla (such as this one) we write proper `exists` function.

Does that make sense to you as well? If so, this should be done in a separate PR

File: src/backend/distributed/commands/utility_hook.c [hunk 24]
```diff
@@ -909,6 +926,43 @@ UndistributeDisconnectedCitusLocalTables(void)
...
+			((AlterObjectSchemaStmt *) node)->objectType == OBJECT_TYPE) ||
+		   (IsA(node, AlterOwnerStmt) && ((AlterOwnerStmt *) node)->objectType ==
+			OBJECT_TYPE) ||
+		   (IsA(node, AlterTableStmt) && AlterTableStmtObjType_compat(
+				((AlterTableStmt *) node)) == OBJECT_TYPE) ||
+		   (IsA(node, DropStmt) && ((DropStmt *) node)->removeType == OBJECT_SEQUENCE) ||
+		   (IsA(node, DropStmt) && ((DropStmt *) node)->removeType ==
+			OBJECT_STATISTIC_EXT) ||
+		   (IsA(node, DropStmt) && ((DropStmt *) node)->removeType == OBJECT_VIEW) ||
+		   (IsA(node, DropStmt) && ((DropStmt *) node)->removeType ==
```
Comment (marcocitus) [inline src/backend/distributed/commands/utility_hook.c:955]: this code is a bit hard to read, is there a way to structure it more?

File: src/backend/distributed/commands/view.c [hunk 25]
```diff
@@ -107,6 +110,32 @@ PostprocessViewStmt(Node *node, const char *queryString)
 		return NIL;
 	}
 
+	if (!DistributeLocalViews)
```
Comment (marcocitus) [inline src/backend/distributed/commands/view.c:113]: was wondering: what problem does this solve? could we maybe add a comment?

File: src/backend/distributed/commands/utility_hook.c [hunk 26]
```diff
@@ -522,13 +523,25 @@ ProcessUtilityInternal(PlannedStmt *pstmt,
 
 	/* only generate worker DDLJobs if propagation is enabled */
 	const DistributeObjectOps *ops = NULL;
+
+	/* determines if we should run preprocess and postprocess steps for the object */
+	bool objValid = true;
```
Comment (marcocitus) [inline src/backend/distributed/commands/utility_hook.c:528]: we usually avoid shorthand / acronyms, so maybe `isObjectValid`

File: src/backend/distributed/commands/type.c [hunk 27]
```diff
@@ -489,7 +523,16 @@ AlterTypeOwnerObjectAddress(Node *node, bool missing_ok)
...
 	TypeName *typeName = makeTypeNameFromNameList((List *) stmt->object);
-	Oid typeOid = LookupTypeNameOid(NULL, typeName, missing_ok);
+	Oid typeOid = LookupTypeNameOid(NULL, typeName, true);
+	if (!OidIsValid(typeOid))
+	{
+		/*
+		 * Citus should not throw error for non-existing objects, let Postgres do that.
+		 * Otherwise, Citus might throw a different error than Postgres, which we don't want.
+		 */
+		return InvalidObjectAddress;
```
Comment (onderkalaci) [inline src/backend/distributed/commands/type.c:533]: there are bunch of test cases that are NOT covered by code-cov. Can you please make sure to cover those?


(Note that maybe we have NOT enabled code-cov for vanilla testing. If not, it'd be great to enable -- though not sure how we enable it).

File: src/backend/distributed/commands/common.c [hunk 28]
```diff
@@ -272,3 +275,71 @@ PreprocessDropDistributedObjectStmt(Node *node, const char *queryString,
...
+		char *tsdictName = NULL;
+		DeconstructQualifiedName(objName, &schemaName, &tsdictName);
+
+		if (!schemaName)
+		{
+			/*
+			 * Citus should not throw error for non-existing objects, let Postgres do that.
+			 * Otherwise, Citus might throw a different error than Postgres, which we don't want.
+			 */
+			Oid tsdictOid = get_ts_dict_oid(objName, missing_ok);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/common.c:304]: what if the `missing_ok` passed here is `false`. Wouldn't we throw error?

lets try to make these functions more generic

File: src/backend/distributed/commands/utility_hook.c [hunk 29]
```diff
@@ -692,7 +711,9 @@ ProcessUtilityInternal(PlannedStmt *pstmt,
 	 */
 	if (EnableDDLPropagation)
 	{
-		if (ops && ops->postprocess)
+		DistObjectCurrentState = DISTOBJECTSTATE_POSTPROCESS;
+
+		if (ops && ops->postprocess && isObjectValid)
```
[the remark this answers is not in the document]
Comment (aykut-bozkurt) [reply src/backend/distributed/commands/utility_hook.c:716]: There is 1 case with 'alter domain xxx drop constraint yyy'. At start, we validate the object calling its address callback. If the constraint is not valid, then it is ok, we are not processing preprocess and postprocess steps. Otherwise, we process those steps. The same address function is called in postprocess step. We end up with an invalid object returned from address callback by mistake because the constraint is already dropped before postprocess. Thus, I need to differentiate those steps. we have just 1 case which poses a problem because we use the same address function in preprocess, postprocess steps as in validation. An alternative solution would be to separate the validation logic from returning address if we want to remove those state enum. Another option would be to use 2 address methods for alterdomain in which we currently call GetObjectAddressFromParseTree in preprocess and postprocess steps. We can have 2 address methods, one of which is used in preprocess and postprocess steps and the other one is used only for validation. If you have other options, we can discuss them as well.

File: src/backend/distributed/commands/utility_hook.c [hunk 30]
```diff
@@ -284,6 +286,8 @@ multi_ProcessUtility(PlannedStmt *pstmt,
 		return;
 	}
 
+	DistObjectCurrentState = DISTOBJECTSTATE_PREPROCESS;
```
[the remark this answers is not in the document]
Comment (aykut-bozkurt) [reply src/backend/distributed/commands/utility_hook.c:289]: I will make it into separate PRs.

File: src/backend/distributed/planner/multi_explain.c [hunk 31]
```diff
@@ -1201,8 +1202,26 @@ CitusExplainOneQuery(Query *query, int cursorOptions, IntoClause *into,
 
 	INSTR_TIME_SET_CURRENT(planstart);
 
-	/* plan the query */
-	PlannedStmt *plan = pg_plan_query_compat(query, NULL, cursorOptions, params);
+	/*
+	 * we should not hide any objects while explaining some query to not break
```
[the remark this answers is not in the document]
Comment (aykut-bozkurt) [reply src/backend/distributed/planner/multi_explain.c:1206]: There are postgres tests which use explain commands querying some pg meta classes. Those tests fail with different plan results because we intercept pg meta class queries and modify the filter.

File: src/backend/distributed/commands/common.c (hunk 28 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/commands/common.c:304]: The same comment is used in other places as well. We should remove it (except a few cases) because we are using the generic missing_ok flag. We pass it as true when we call address methods.

File: src/backend/distributed/commands/utility_hook.c (hunk 30 above)
[the remark this answers is not in the document]
Comment (aykut-bozkurt) [reply src/backend/distributed/commands/utility_hook.c:289]: Explained below.

File: src/backend/distributed/commands/utility_hook.c (hunk 29 above)
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/commands/utility_hook.c:716]: > . The same address function is called in postprocess step

I think this should be a generic problem.

Should we do the following change such that we do not care about `ALlTER .. DROP ` cases? Would be good to have a separate PR if that solves your problem

```
diff --git a/src/backend/distributed/commands/common.c b/src/backend/distributed/commands/common.c
index 1c6a71de3..71d53f0ac 100644
--- a/src/backend/distributed/commands/common.c
+++ b/src/backend/distributed/commands/common.c
@@ -156,7 +156,7 @@ PostprocessAlterDistributedObjectStmt(Node *stmt, const char *queryString)
        const DistributeObjectOps *ops = GetDistributeObjectOps(stmt);
        Assert(ops != NULL);
 
-       ObjectAddress address = GetObjectAddressFromParseTree(stmt, false);
+       ObjectAddress address = GetObjectAddressFromParseTree(stmt, true);
        if (!ShouldPropagateObject(&address))
        {
                return NIL;

```

**Comment by aykut-bozkurt** [comment]
I have divided this PR into sub PRs because we are doing many sub tasks here and it is getting harder to review.

1. #6055,
2. #6056,
3. Use address callback to verify distributed object validity,
4. EnablePropagationWarning GUC related changes,
5. ResolvePctType GUC related changes,
6. DistributeLocalView GUC related changes,
7. Changes related to drop constraint for alter domain.

**Comment by onderkalaci** [comment]
as we discussed, lets make sure check-vanilla passed on PG 15 as well, on #6085. Noting for broader visibility & future reference

File: src/test/regress/pg_regress_multi.pl [hunk 32]
```diff
@@ -1019,21 +1021,19 @@ END
...
+    # after dropping it if exists. Thus, we set use-existing flag and
+    # manually create the default db, citus extension and is_citus_depended_object
+    # function. When use-existing flag is set, pg_regress does not drop
+    # default db.
     ###
 
-    my $dbName = shift;
-
-    # prepare tablespace folder
-    rmdir "./testtablespace";
```
Comment (onderkalaci) [inline src/test/regress/pg_regress_multi.pl:1035]: Hmm, on my local machine tablespace test is failing. I wonder if it is related to removing these?

**Review by onderkalaci — APPROVED** [review]
Yay!

=======================
 All 210 tests passed. 
=======================

File: .circleci/config.yml [hunk 33]
```diff
@@ -413,6 +413,76 @@ jobs:
           flags: 'test_<< parameters.pg_major >>,<< parameters.make >>'
           when: always
 
+  test-pg-vanilla:
```
[the remark this answers is not in the document]
Comment (aykut-bozkurt) [reply .circleci/config.yml:416]: We have tests with the same name as pg's tests, so we have conflict on those files during tests. We simply git ignore the output folder for pg tests.

File: src/test/regress/pg_regress_multi.pl [hunk 34]
```diff
@@ -1019,21 +1021,19 @@ END
...
-    mkdir "./testtablespace";
+    $ENV{VANILLATEST} = "1";
```
[the remark this answers is not in the document]
Comment (aykut-bozkurt) [reply src/test/regress/pg_regress_multi.pl:1035]: It is used in src/test/regress/bin/diff.sh and it was set before as well during pg vanilla tests.

File: src/test/regress/pg_regress_multi.pl [hunk 35]
```diff
@@ -1068,40 +1063,45 @@ sub PrepForVanillaTest
...
-                ("--use-existing"),
-                ("--host","$host"),
-                ("--port","$masterPort"),
-                ("--user","$user"),
-                ("--dbname", "$dbName"))
+	    $pgregressdir=catfile("$postgresSrcdir", "src", "test", "regress");
 	}
+
+    # output dir
+    my $pgregressOutputdir = "$citusAbsSrcdir/pg_vanilla_outputs/$majorversion";
```
[the remark this answers is not in the document]
Comment (aykut-bozkurt) [reply src/test/regress/pg_regress_multi.pl:1078]: We always recreate all output folders before vanilla tests run.

File: .circleci/config.yml (hunk 33 above)
[the remark this answers is not in the document]
Comment (aykut-bozkurt) [reply .circleci/config.yml]: I removed it. We no longer need it because I give symlink to regression.diff to be found in the same place with other tests.
